### ISSUE TITLE
[Question]Is there any way to select a sample from the discrete candidate space with acquisition functions?

### ISSUE THREAD

**Issue by Waxpple** [pytorch/botorch#1137]
## Issue description
1. Is there any way to select a sample from the discrete candidate space with acquisition functions?
I want to get new samples from a dataset `X_cand`, which is a discrete space.
Thompson sampling chooses the sample with maxima posterior. 
If I want to use EI as my acquisition function, and I use the Boltzmann sampling function with `eta=0`, would it be the same as I want? Or there is any way to do this discrete sampling?
2. Should I use torch.no_grad() on Boltzmann?
## Code example
```python
X_cand = [[0,0,0], [1,1,3]...] # a discrete dataset
acq = ExpectedImprovement(model, train_Y.max(), maximize=True)
boltzmann = BoltzmannSampling(acq_func=acq,eta=0.0,replacement=False)

with torch.no_grad():
    X_next = boltzmann(X_cand, num_samples=batch_size)
```

## System Info
- BoTorch Version 0.6.1
- GPyTorch Version 1.6.0
- PyTorch Version 1.10.1+cu113
- Ubuntu 18.04

Thanks!

Best regards,
Waxpple

**Comment by Balandat** [comment]
You can use `optimize_acqf_discrete` for that.

**Comment by Waxpple** [comment]
> You can use `optimize_acqf_discrete` for that.

Great, Thank you! And one question, when do I know should use `torch.no_grad()` or not?

**Comment by saitcakmak** [comment]
> when do I know should use torch.no_grad() or not?

You should use it when you don't need gradients from an operation that traditionally produces gradients. This could be sampling from the GP posterior via `rsample` (which produces gradients if the inputs `X` have `requires_grad=True`) for producing some plots. On the other hand, if I were interested in computing some quantities (e.g., some acquisition function value) from the output of `rsample` and use that to optimize `X`, then I wouldn't use `no_grad` and I'd make sure that `X.requires_grad=True` (our `optimize_acqf` and similar helpers do this internally).

Just a plug regarding your original question: If you're interested in optimizing some function over a discrete space using BO, you may find it easier to use [Ax](ax.dev) (which uses BoTorch under the hood). We also have [tutorials](https://botorch.org/tutorials/custom_botorch_model_in_ax) showing how to use a custom BoTorch model or acquisition function with Ax, in case you're interested in developing your own.

**Comment by Waxpple** [comment]
I got it now thank you!
Yes, I will love to try using `Ax`. I just want to figure `BoTorch` out since it is the backbone of Ax.

Thanks
